### PR TITLE
Added option to report metrics to kafka for gobblin runs.

### PR DESCRIPTION
**Pull request by ibuenros**

### CODE REVIEW

**Comment by ibuenros** [comment]
@liyinan926 @zliu41 can you review?

File: gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java [hunk 1]
```diff
@@ -238,4 +253,42 @@ private void buildJmxMetricReporter(Properties properties) {
...
+    if (!Boolean.valueOf(properties.getProperty(ConfigurationKeys.METRICS_REPORTING_KAFKA_ENABLED_KEY,
+        ConfigurationKeys.DEFAULT_METRICS_REPORTING_KAFKA_ENABLED))) {
+      LOGGER.info("Not reporting metrics to Kafka");
+      return;
+    }
+
+    String brokers = properties.getProperty(ConfigurationKeys.METRICS_KAFKA_BROKERS);
+    String topic = properties.getProperty(ConfigurationKeys.METRICS_KAFKA_TOPIC);
+
+    if (Strings.isNullOrEmpty(brokers)) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java]: You can use `Preconditions.checkArgument(properties.containsKey(...))` to check the presence of those two properties.

File: gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java [hunk 2]
```diff
@@ -238,4 +253,42 @@ private void buildJmxMetricReporter(Properties properties) {
...
+
+    if (Strings.isNullOrEmpty(topic)) {
+      LOGGER.error("Kafka metrics topic property is empty or not supplied. Will not report metrics to Kafka.");
+      return;
+    }
+
+    String reportingFormat = properties.getProperty(ConfigurationKeys.METRICS_REPORTING_KAFKA_FORMAT,
+        ConfigurationKeys.DEFAULT_METRICS_REPORTING_KAFKA_FORMAT);
+
+    if (!reportingFormat.equalsIgnoreCase("avro") && !reportingFormat.equalsIgnoreCase("json")) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/metrics/GobblinMetrics.java]: Can we define supported reporting format as a `enum`?

**Comment by ibuenros** [comment]
Addressed comments.

**Comment by ibuenros** [comment]
Addressed.

**Comment by zliu41** [comment]
LGTM

File: gobblin-metrics/src/main/java/gobblin/metrics/kafka/KafkaReportingFormats.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  public KafkaReporter.Builder<?> reporterBuilder(MetricContext context) {
+    switch (this) {
+      case AVRO:
+        return KafkaAvroReporter.forContext(context);
+      case JSON:
+        return KafkaReporter.forContext(context);
+      default:
+        // This should never happen.
+        throw new RuntimeException("KafkaReportingFormat not recognized.");
```
Comment (sahilTakiar) [inline gobblin-metrics/src/main/java/gobblin/metrics/kafka/KafkaReportingFormats.java]: Why not make this an IllegalStateException or IllegalArgumentException

File: gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java [hunk 4]
```diff
@@ -127,7 +128,11 @@ public T outputTo(PrintStream output) {
      * @return {@code this}
      */
     public T outputTo(OutputStream stream) {
-      this.output = new PrintStream(stream);
+      try {
+        this.output = new PrintStream(stream, false, "UTF-8");
```
Comment (sahilTakiar) [inline gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java]: You can use Guava's Charsets.UTF-8 to be consistent with the rest of the Gobblin code.

File: gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java [hunk 5]
```diff
@@ -127,7 +128,11 @@ public T outputTo(PrintStream output) {
...
+      } catch(UnsupportedEncodingException exception) {
+        LOGGER.error("Unsupported encoding in OutputStreamReporter. This is an error with the code itself.", exception);
```
Comment (sahilTakiar) [inline gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java]: Any reason not to rethrow this exception?

File: gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java (hunk 4 above)
Comment (liyinan926) [inline gobblin-metrics/src/main/java/gobblin/metrics/OutputStreamReporter.java]: Yeah, we do have `ConfigurationKeys.DEFAULT_CHARSET_ENCODING`.

**Comment by ibuenros** [comment]
@sahilTakiar addressed comments.

**Comment by sahilTakiar** [comment]
LGTM

**Pull request merged**